File: task3/main.py
import os
from fastapi import FastAPI, HTTPException
from pydantic import BaseModel, Field, field_validator
from typing import Optional

app = FastAPI(title="Task3 API", version="1.0.0")

items_db: dict = {}
next_id: int = 1


class Item(BaseModel):
    id: Optional[int] = None
    name: str
    description: Optional[str] = None
    price: float = Field(..., gt=0)

    @field_validator("name")
    @classmethod
    def name_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("name cannot be empty or whitespace")
        return v


class ItemResponse(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    price: float


class EchoRequest(BaseModel):
    message: str


class EchoResponse(BaseModel):
    message: str
    length: int


@app.get("/")
async def root():
    return {"message": "Welcome to Task3 API"}


@app.get("/health")
async def health_check():
    return {"status": "healthy"}


@app.post("/echo", response_model=EchoResponse)
async def echo(request: EchoRequest):
    return EchoResponse(message=request.message, length=len(request.message))


@app.get("/items", response_model=list[ItemResponse])
async def get_items():
    return [
        ItemResponse(id=item.id, name=item.name, description=item.description, price=item.price)
        for item in items_db.values()
    ]


@app.post("/items", response_model=ItemResponse, status_code=201)
async def create_item(item: Item):
    global next_id
    new_id = next_id
    next_id += 1
    new_item = Item(id=new_id, name=item.name, description=item.description, price=item.price)
    items_db[new_id] = new_item
    return ItemResponse(id=new_id, name=item.name, description=item.description, price=item.price)


@app.get("/items/{item_id}", response_model=ItemResponse)
async def get_item(item_id: int):
    if item_id not in items_db:
        raise HTTPException(status_code=404, detail="Item not found")
    item = items_db[item_id]
    return ItemResponse(id=item.id, name=item.name, description=item.description, price=item.price)


@app.put("/items/{item_id}", response_model=ItemResponse)
async def update_item(item_id: int, item: Item):
    if item_id not in items_db:
        raise HTTPException(status_code=404, detail="Item not found")
    updated_item = Item(id=item_id, name=item.name, description=item.description, price=item.price)
    items_db[item_id] = updated_item
    return ItemResponse(id=updated_item.id, name=updated_item.name, description=updated_item.description, price=updated_item.price)


@app.delete("/items/{item_id}", status_code=204)
async def delete_item(item_id: int):
    if item_id not in items_db:
        raise HTTPException(status_code=404, detail="Item not found")
    del items_db[item_id]


if __name__ == "__main__":
    import uvicorn
    port = int(os.getenv("PORT", 8000))
    uvicorn.run(app, host="0.0.0.0", port=port)

File: task3/main.go
package main

import (
	"fmt"
	"net/http"
	"os"
	"sync"

	"github.com/gin-gonic/gin"
)

type Item struct {
	ID          *int     `json:"id"`
	Name        string   `json:"name"`
	Description *string  `json:"description"`
	Price       float64  `json:"price"`
}

type ItemResponse struct {
	ID          int     `json:"id"`
	Name        string  `json:"name"`
	Description *string `json:"description"`
	Price       float64 `json:"price"`
}

type EchoRequest struct {
	Message string `json:"message"`
}

type EchoResponse struct {
	Message string `json:"message"`
	Length  int    `json:"length"`
}

var (
	itemsDB  = make(map[int]Item)
	nextID   = 1
	itemsMux sync.RWMutex
)

func main() {
	r := gin.Default()

	r.GET("/", func(c *gin.Context) {
		c.JSON(http.StatusOK, gin.H{
			"message": "Welcome to Task3 API",
		})
	})

	r.GET("/health", func(c *gin.Context) {
		c.JSON(http.StatusOK, gin.H{
			"status": "healthy",
		})
	})

	r.POST("/echo", func(c *gin.Context) {
		var req EchoRequest
		if err := c.ShouldBindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
			return
		}
		c.JSON(http.StatusOK, EchoResponse{
			Message: req.Message,
			Length:  len(req.Message),
		})
	})

	r.GET("/items", func(c *gin.Context) {
		itemsMux.RLock()
		defer itemsMux.RUnlock()

		result := make([]ItemResponse, 0, len(itemsDB))
		for _, item := range itemsDB {
			result = append(result, ItemResponse{
				ID:          *item.ID,
				Name:        item.Name,
				Description: item.Description,
				Price:       item.Price,
			})
		}
		c.JSON(http.StatusOK, result)
	})

	r.POST("/items", func(c *gin.Context) {
		var item Item
		if err := c.ShouldBindJSON(&item); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
			return
		}

		itemsMux.Lock()
		newID := nextID
		nextID++
		newItem := Item{
			ID:          &newID,
			Name:        item.Name,
			Description: item.Description,
			Price:       item.Price,
		}
		itemsDB[newID] = newItem
		itemsMux.Unlock()

		c.JSON(http.StatusCreated, ItemResponse{
			ID:          *newItem.ID,
			Name:        newItem.Name,
			Description: newItem.Description,
			Price:       newItem.Price,
		})
	})

	r.GET("/items/:id", func(c *gin.Context) {
		id, err := parseParam(c, "id")
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid item ID"})
			return
		}

		itemsMux.RLock()
		item, exists := itemsDB[id]
		itemsMux.RUnlock()

		if !exists {
			c.JSON(http.StatusNotFound, gin.H{"error": "Item not found"})
			return
		}

		c.JSON(http.StatusOK, ItemResponse{
			ID:          *item.ID,
			Name:        item.Name,
			Description: item.Description,
			Price:       item.Price,
		})
	})

	r.PUT("/items/:id", func(c *gin.Context) {
		id, err := parseParam(c, "id")
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid item ID"})
			return
		}

		var item Item
		if err := c.ShouldBindJSON(&item); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
			return
		}

		itemsMux.Lock()
		_, exists := itemsDB[id]
		if !exists {
			itemsMux.Unlock()
			c.JSON(http.StatusNotFound, gin.H{"error": "Item not found"})
			return
		}

		updatedItem := Item{
			ID:          &id,
			Name:        item.Name,
			Description: item.Description,
			Price:       item.Price,
		}
		itemsDB[id] = updatedItem
		itemsMux.Unlock()

		c.JSON(http.StatusOK, ItemResponse{
			ID:          *updatedItem.ID,
			Name:        updatedItem.Name,
			Description: updatedItem.Description,
			Price:       updatedItem.Price,
		})
	})

	r.DELETE("/items/:id", func(c *gin.Context) {
		id, err := parseParam(c, "id")
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid item ID"})
			return
		}

		itemsMux.Lock()
		_, exists := itemsDB[id]
		if !exists {
			itemsMux.Unlock()
			c.JSON(http.StatusNotFound, gin.H{"error": "Item not found"})
			return
		}
		delete(itemsDB, id)
		itemsMux.Unlock()

		c.Status(http.StatusNoContent)
	})

	port := os.Getenv("PORT")
	if port == "" {
		port = "8080"
	}
	r.Run(":" + port)
}

func parseParam(c *gin.Context, key string) (int, error) {
	var id int
	_, err := fmt.Sscanf(c.Param(key), "%d", &id)
	return id, err
}
